server/plugin/client.go

```diff
@@ -0,0 +1,549 @@
+				}
+
+				c.plugin.API.PublishWebSocketEvent(
+					constants.WSEventDisconnect,
+					nil,
+					&model.WebsocketBroadcast{UserId: mattermostUserID},
+				)
+
+				if _, DMErr := c.plugin.DM(mattermostUserID, message, false); DMErr != nil {
+					c.plugin.API.LogError(constants.UnableToDMBot, "Error", DMErr.Error())
+				}
+				return nil, http.StatusInternalServerError, errRefreshingToken
+			}
+		}
+	}
+
+	var req *http.Request
+	if formValues != nil {
+		req, err = http.NewRequest(method, URL, strings.NewReader(formValues.Encode()))
+		if err != nil {
+			return nil, http.StatusInternalServerError, err
+		}
+	} else {
+		req, err = http.NewRequest(method, URL, inBody)
+		if err != nil {
+			return nil, http.StatusInternalServerError, err
+		}
+	}
+
+	if mattermostUserID != "" {
+		if err = c.plugin.AddAuthorization(req, mattermostUserID); err != nil {
+			return nil, http.StatusInternalServerError, err
+		}
+	}
+
+	return c.makeHTTPRequest(req, contentType, out)
+}
+
+func (c *client) OpenDialogRequest(body *model.OpenDialogRequest, mattermostUserID string) (int, error) {
+	_, statusCode, err := c.CallJSON(c.plugin.getConfiguration().MattermostSiteURL, constants.PathOpenCommentModal, http.MethodPost, mattermostUserID, body, nil, nil)
+	return statusCode, err
+}
+
+func (c *client) parsePath(basePath, path, method string) (string, error) {
+	pathURL, err := url.Parse(path)
+	if err != nil {
+		return "", err
+	}
+
+	if pathURL.Scheme == "" || pathURL.Host == "" {
+		var baseURL *url.URL
+		baseURL, err = url.Parse(basePath)
+		if err != nil {
+			return "", err
+		}
+		if path[0] != '/' {
+			path = "/" + path
+		}
+		path = baseURL.String() + path
+	}
+
+	return path, nil
+}
+
+func (c *client) makeHTTPRequest(req *http.Request, contentType string, out interface{}) (responseData []byte, statusCode int, err error) {
+	if contentType != "" {
+		req.Header.Add("Content-Type", contentType)
+	}
+
+	resp, err := c.httpClient.Do(req)
+	if err != nil {
+		return nil, http.StatusInternalServerError, err
+	}
+
+	if resp.Body == nil {
+		return nil, resp.StatusCode, nil
+	}
+	defer resp.Body.Close()
+
+	responseData, err = io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, http.StatusInternalServerError, err
+	}
+
+	switch resp.StatusCode {
+	case http.StatusOK, http.StatusCreated:
+		if out != nil {
+			if err = json.Unmarshal(responseData, out); err != nil {
+				return responseData, http.StatusInternalServerError, err
+			}
+		}
+		return responseData, resp.StatusCode, nil
+
+	case http.StatusNoContent:
+		return nil, resp.StatusCode, nil
+
+	case http.StatusNotFound:
+		return nil, resp.StatusCode, ErrNotFound
+	}
+
+	errResp := ErrorResponse{}
+	if err = json.Unmarshal(responseData, &errResp); err != nil {
+		return responseData, http.StatusInternalServerError, errors.WithMessagef(err, "status: %s", resp.Status)
+	}
+	return responseData, resp.StatusCode, fmt.Errorf("errorMessage %s", errResp.Message)
```

Comment:
For this type:

```go
type ErrorResponse struct {
	Message string `json:"message"`
}
```

As long as the response is valid JSON data, and it doesn't contain a `message` field that has a type other than `string`, this if block won't ever run. In other words, `message` is not necessarily defined when the marshal error does not occur. I'm thinking we should log the response code here since we potentially don't have an error message to use in this case.

Reply:
Bringing this up because there was an OAuth error @DHaussermann ran into when there was a configuration error for the app on Azure's side. The error message was blank so it read as: `errorMessage` with no message afterwards.